src/centroiders.cpp

```diff
@@ -25,6 +25,88 @@ std::vector<Star> DummyCentroidAlgorithm::Go(unsigned char *image, int imageWidt
     return result;
 }
 
+// a poorly designed thresholding algorithm
+int BadThreshold(unsigned char *image, int imageWidth, int imageHeight) {
+    //loop through entire array, find sum of magnitudes
+    int totalMag = 0;
+    for (int i = 0; i < imageHeight * imageWidth; i++) {
+        totalMag += image[i];
+    }
+    return (((totalMag/(imageHeight * imageWidth)) + 1) * 15) / 10;
+}
+
+// a more sophisticated thresholding algorithm, not tailored to star images
+int OtsusThreshold(unsigned char *image, int imageWidth, int imageHeight) {
+    // code here, duh
+    float total = imageWidth * imageHeight;
+    //float top = 255;
+    float sumB = 0;
+    float sum1 = 0;
+    float wB = 0;
+    float maximum = 0;
+    int level = 0;
+    // make the histogram (array length 256)
+    int * histogram = new int[256];
+    for (int i = 0; i < 256; i++) {
+        histogram[i] = 0;
+    }
+    for (int i = 0; i < total; i++) {
+        histogram[image[i]] ++;
+    }
+    for (int i = 0; i < 256; i ++) {
+        sum1 += i * histogram[i];
+    }
+    for (int i = 0; i < 256; i ++) {
+        float wF = total - wB;
+        //std::cout << "wF\n" << wB << "\n";
+        //std::cout << "wB\n" << wF << "\n";
+        if (wB > 0 && wF > 0) {
+            float mF = (sum1 - sumB) / wF;
+            float val = wB * wF * ((sumB / wB) - mF) * ((sumB / wB) - mF);
+            //std::cout << val << "\n";
+            if (val >= maximum) {
+                level = i;
+                maximum = val;
+            }
+        }
+        wB = wB + histogram[i];
+        sumB = sumB + i * histogram[i];
+    }
+    return level;
+}
+
+// a simple, but well tested thresholding algorithm that works well with star images
+int BasicThreshold(unsigned char *image, int imageWidth, int imageHeight) {
+    int totalMag = 0;
+    float std = 0;
+    int totalPixels = imageHeight * imageWidth;
```

Comment:
`long`, even with a 1MP image this will be about 1,000,000 which is not guaranteed to fit in an int.